src/forces.jl

```diff
@@ -0,0 +1,81 @@
+"""
+Computes minus the derivatives of the energy with respect to atomic positions.
+"""
+function forces(scfres)
+    # By generalized Hellmann-Feynman, dE/dR = ∂E/∂R. The atomic
+    # positions come up explicitly only in the external and nonlocal
+    # part of the energy,
+
+    # TODO this assumes that the build_external and build_nonlocal are consistent with model.atoms
+    # Find a way to move this computation closer to each term
+
+    # minus signs here because f = -∇E
+    ham = scfres.ham
+    basis = ham.basis
+    model = basis.model
+    Psi = scfres.Psi
+    ρ = scfres.ρ
+    T = real(eltype(ham))
+
+    forces = []
+    for (type, pos) in model.atoms
+        @assert type.psp != nothing
+        f = zeros(Vec3{T}, length(pos))
+
+        # Local part
+        # energy = sum of form_factor(G) * struct_factor(G) * rho(G)
```

Comment:
Is form_factor the right terminology here? Isn't it just the FT of the local part of the PSP?

Reply:
At least you should not use the same terminology between this line and line 48 (where "form factor" is coined as the FT of the non-local part).

Reply:
I think yes, this is standard terminology in crystallography/scattering/all this physics I don't really undertsand. Form factor is the info for an atom at 0, structure factor is the phase factors coming from the positions of the atoms. See eg https://en.wikipedia.org/wiki/Structure_factor https://en.wikipedia.org/wiki/Atomic_form_factor

Reply:
Ah I see so form factor is just a fancy word for fourier coefficients.

Reply:
Everything is fancy words for Fourier :-p I think it's to distinguish between the characteristics of individual atoms and of the crystal structure, it's a good analogy to use

Reply:
> Everything is fancy words for Fourier :-p

Words of a true French mathematician :).